### PR TITLE
Feature: Item-specific Stack Sizes [Part 1/10 of Item Stack Size (formerly #694, formerly restoring #468)]

### PR DESCRIPTION
**Pull request by RayDeeUx**
hanni wanted me to split #694 so here we are. ignore the commit count, this is the only way i could split things without cut and paste abuse

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/features/inventory/ItemStars.kt
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/ItemStars.kt:1]: The change from `CRIMSON_ARMOR.isSelected()` to     `!SkyHanniMod.feature.inventory.stackSize.itemNumber.contains(StackSizeConfig.ItemNumber.CRIMSON_STARS)` does feel like a downgrade

File: src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt [hunk 1]
```diff
@@ -259,6 +260,12 @@ object LorenzUtils {
 
     fun getPlayerName(): String = Minecraft.getMinecraft().thePlayer.name
 
+    fun isRewardChest(): Boolean {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt:287]: What kind of reward chest is this? either belogs in invenotry utils or inside cata api

File: src/main/java/at/hannibal2/skyhanni/features/inventory/ItemStars.kt
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/ItemStars.kt:1]: quoting directly from my own discord msgs (which, yes, that means missing punctuation and occasionally poor phrasing):
> it only feels like a downgrade because isselected relies on importing individual members of the enum, and with your ban on star imports id rather be consistent across all the class files while saving on number of import lines
> tldr isselected is wack id rather just do old-fashioned .contains() for consistency

File: src/main/java/at/hannibal2/skyhanni/utils/SkyBlockItemModifierUtils.kt [hunk 2]
```diff
@@ -181,7 +196,29 @@ object SkyBlockItemModifierUtils {
...
     fun ItemStack.getBottleOfJyrreSeconds() = getAttributeInt("bottle_of_jyrre_seconds")
 
-    fun ItemStack.getEdition() = getAttributeInt("edition")
+    fun ItemStack.getPrehistoricEggBlocksWalked() = getAttributeInt("blocks_walked")
+
+    fun ItemStack.getNecronHandlesFound() = getAttributeInt("handles_found")
+
+    fun ItemStack.getBloodGodKills() = getAttributeInt("blood_god_kills")
+    
+    fun ItemStack.getYetiRodFishesCaught() = getAttributeInt("fishes_caught")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/SkyBlockItemModifierUtils.kt]: Are we sure this stat is only on the yeti rod?

File: src/main/java/at/hannibal2/skyhanni/utils/CachedItemData.kt [hunk 3]
```diff
@@ -21,4 +21,6 @@ data class CachedItemData(
 
     // null = not loaded
     var itemRarity: LorenzRarity? = null,
+    /** -1 = not loaded */ var auctionNumber: Int? = -1,
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/CachedItemData.kt]: please use the same formatting as in the rest of the file

File: src/main/java/at/hannibal2/skyhanni/utils/SkyBlockItemModifierUtils.kt [hunk 4]
```diff
@@ -181,7 +196,29 @@ object SkyBlockItemModifierUtils {
...
+
+    fun ItemStack.getEdition(): Int? {
+        val data = cachedData
+        if (data.editionNumber == -1) {
+            data.editionNumber = getAttributeInt("edition")
+        }
+        return data.editionNumber
+    }// = getAttributeInt("edition")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/SkyBlockItemModifierUtils.kt]: Those 2 comments can go

File: src/main/java/at/hannibal2/skyhanni/SkyHanniMod.kt [hunk 5]
```diff
@@ -427,7 +427,7 @@ class SkyHanniMod {
         loadModule(PlayerChatModifier())
         loadModule(DungeonChatFilter())
         loadModule(HideNotClickableItems())
-        loadModule(ItemDisplayOverlayFeatures)
+        loadModule(ItemDisplayOverlayFeatures())
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/SkyHanniMod.kt]: Any particular reason why the object got changed back to a class?

File: src/main/java/at/hannibal2/skyhanni/features/inventory/itemdisplayoverlay/ItemDisplayOverlayFeatures.kt [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+import at.hannibal2.skyhanni.utils.SkyBlockItemModifierUtils.getFruitBowlNames
+import at.hannibal2.skyhanni.utils.SkyBlockItemModifierUtils.getNecronHandlesFound
+import at.hannibal2.skyhanni.utils.SkyBlockItemModifierUtils.getPrehistoricEggBlocksWalked
+import at.hannibal2.skyhanni.utils.SkyBlockItemModifierUtils.getYetiRodFishesCaught
+import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+import kotlin.math.floor
+import kotlin.math.log10
+
+class ItemDisplayOverlayFeatures {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/itemdisplayoverlay/ItemDisplayOverlayFeatures.kt]: Can you make this class not "new" but rather "moved from to"? Otherwise we will lose the whole git history of the file.

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+public class StackSizeConfig {
+
+    @Expose
+    @ConfigOption(
+        name = "Item Number",
+        desc = "Showing the item number as a stack size for these items." // Some values may be truncated percentages or §a✔§r§7s.
+    )
+    @ConfigEditorDraggableList
+    public List<ItemNumber> itemNumber = new ArrayList<>(Arrays.asList(
+        ItemNumber.MASTER_STAR,
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java]: This list does not get migrated correctly. I just have the default things enabled when testing with a legacy config. This should not happen

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        ItemNumber.SKYBLOCK_YEAR,
+        ItemNumber.CRIMSON_STARS,
+        ItemNumber.LARVA_HOOK,
+        ItemNumber.ARMADILLO,
+        ItemNumber.BEASTMASTER,
+        ItemNumber.CAMPFIRE
+    ));
+
+    public enum ItemNumber implements HasLegacyId {
+        MASTER_STAR("§bMaster Star Tier", 0),
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java]: Why are you changing almost all the names of the enums?

File: src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        CRIMSON_STARS("§bCrimson Armor Crimson Stars", 6),
+        CAMPFIRE("§bCampfire Talisman Tier", 7),
+        KUUDRA("§bKuudra Key", 8),
+        SOULFLOW("§bInternalized Soulflow Count\n§b(Abbv, won't show in the Auction House)", 9),
+        BLOOD_GOD("§bBlood God Crest Strength", 10),
+        RANCHER_SPEED("§bRancher's Boots Speed", 11),
+        LARVA_HOOK("§bLarva Hook", 12),
+        DUNGEON_POTION_LEVEL("§bDungeon Potion Level", 13),
+        VACCUM_PESTS("§bPests Stored in Vaccum (Garden)", 14),
+        JYRRE("§bBottle of Jyrre (#, Intelligence Bonus on Legacy Bottles and Time on New Bottles)", 15),
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/inventory/stacksize/StackSizeConfig.java]: name too long, can be removed entirely. time on new bottles has nothing to do with this feature, should be it's own enum instance

**Review by hannibal002 — CHANGES_REQUESTED** [review]
The history of a file you moved gets lost entirely, the whole item stack list gets reset in the config, and many lines aren't readable at all. You can add the additional information about the enum instances as comments so we can include it once we have a "hover over to view description" logic.

https://github.com/hannibal002/SkyHanni/pull/755 will be merged first. You might want to wait for this to happen before changing stuff in `ItemDisplayOverlayFeatures`. Or merge walkers changes into your PR, so you can start changing the file right now.

File: src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt (hunk 1 above)
Comment (RayDeeUx) [reply src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt:263]: https://discord.com/channels/997079228510117908/1153406580621066301/1155039211674685450

**Review by CalMWolfs — CHANGES_REQUESTED** [review]
ok so much better than i thought it would be (sorry) but my main problem is with the names of variables and functions. While descriptive they seem too long (and too descriptive) which makes the code less readable.
also the base key group along with the individual pattern keys should also be shortened a bit but also split into more words split by fullstops instead of it being like 2 sets of 3-4 words separated by a fullstop

**Comment by RayDeeUx** [comment]
> but my main problem is with the names of variables and functions. 

i'm content with renaming the variables, but for functions, a good chunk of it is modeled off of @walkerselby's PR #755:
```kt
return when {
    isMasterStar(internalName) -> getMasterStarTip(internalName)
    isMasterSkull(internalName) -> getMasterSkullTip(internalName)
    isDungeonHead(internalName) -> getDungeonHeadTip(internalName)
    isNewYearCake(internalName) -> getNewYearCakeTip(item)
    isPet(itemName) -> getPetTip(itemName)
    isMinionTier(itemName, item) -> getMinionTierTip(itemName)
    isSack(item) -> getSackTip(itemName)
    isKuudraKey(internalName) -> getKuudraKeyTip(internalName)
    isSkillLevel(item) -> getSkillLevelTip(item, itemName)
    isCollectionLevel() -> getCollectionLevelTip(item)
    isRanchersBoots(internalName) -> getRanchersBootsTip(item)
    isLarvaHook(internalName) -> getLarvaHookTip(item)
    isDungeonPotion(item, internalName) -> getDungeonPotionTip(item)
    isVacuumGarden(item) -> getVacuumGardenTip(item)
    isBottleOfJyrre(item) -> getBottleOfJyrreTip(item)
    isEditionNumber(item) -> getEditionNumberTip(item)
    isBingoGoalRank(chestName, item) -> getBingoGoalRankTip(item)
    else -> ""
}
```
this particular PR is supposed to carry that torch, so to speak. if you still want me to rename the functions i can start with removing the `Tip` suffix from most of the functions

also side note: i have't gotten around to using the `return when` statement like walker did for PRs #760 through #768

**Comment by CalMWolfs** [comment]
Looked at the function names and while they are long it's ok but please do change the variable names
